### PR TITLE
dhcp: further cleanups for timestamp

### PR DESCRIPTION
**Pull request by yuwata**
Follow-ups for #29910.

### CODE REVIEW

File: src/libsystemd-network/sd-dhcp-client.c [hunk 1]
```diff
@@ -2012,13 +2012,64 @@ static int client_restart(sd_dhcp_client *client) {
...
+                return log_dhcp_client_errno(client, SYNTHETIC_ERRNO(EBADMSG),
+                                             "Too small to be a DHCP message: ignoring.");
+
+        if (be32toh(message->magic) != DHCP_MAGIC_COOKIE)
+                return log_dhcp_client_errno(client, SYNTHETIC_ERRNO(EBADMSG),
+                                             "Not a DHCP message: ignoring.");
+
+        if (message->op != BOOTREPLY)
+                return log_dhcp_client_errno(client, SYNTHETIC_ERRNO(EBADMSG),
+                                             "Not a BOOTREPLY message: ignoring.");
```
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp-client.c]: ", ignoring".

File: src/libsystemd-network/sd-dhcp-client.c [hunk 2]
```diff
@@ -2012,13 +2012,64 @@ static int client_restart(sd_dhcp_client *client) {
...
+        }
+
+        if (message->hlen != expected_hlen)
+                return log_dhcp_client_errno(client, SYNTHETIC_ERRNO(EBADMSG),
+                                             "Received packet hlen (%u) does not match expected (%u), ignoring.",
+                                             message->hlen, expected_hlen);
+
+        if (memcmp_safe(message->chaddr, expected_chaddr, expected_hlen))
+                return log_dhcp_client_errno(client, SYNTHETIC_ERRNO(EBADMSG),
+                                             "Received chaddr does not match expected: ignoring.");
```
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp-client.c]: Also here.

File: src/libsystemd-network/dhcp-server-internal.h [hunk 3]
```diff
@@ -110,12 +110,13 @@ typedef struct DHCPRequest {
...
         size_t parameter_request_list_len;
         bool rapid_commit;
+        triple_timestamp timestamp;
 } DHCPRequest;
 
 extern const struct hash_ops dhcp_lease_hash_ops;
 
 int dhcp_server_handle_message(sd_dhcp_server *server, DHCPMessage *message,
-                               size_t length);
+                               size_t length, const triple_timestamp *t);
```
Comment (keszybz) [inline src/libsystemd-network/dhcp-server-internal.h]: "timestamp" for clarity?

File: src/libsystemd-network/sd-dhcp-server.c [hunk 4]
```diff
@@ -897,6 +897,31 @@ static int ensure_sane_request(sd_dhcp_server *server, DHCPRequest *req, DHCPMes
         return 0;
 }
 
+static void request_set_timestamp(DHCPRequest *req, const triple_timestamp *timestamp) {
+        assert(req);
+
+        if (timestamp && triple_timestamp_is_set(timestamp))
+                req->timestamp = *timestamp;
+        else
+                triple_timestamp_get(&req->timestamp);
```
Comment (keszybz) [inline src/libsystemd-network/sd-dhcp-server.c:906]: I think the naming of the `*_timestamp_get()` functions is confusing. But that's a separate topic. It's called correctly here.

**Review by keszybz — COMMENTED** [review]
LGTM.

File: src/libsystemd-network/sd-dhcp-server.c (hunk 4 above)
Comment (keszybz) [reply src/libsystemd-network/sd-dhcp-server.c:906]: I prepared a patch to change this. I'll submit it once this PR is merged.

**Comment by yuwata** [comment]
@keszybz Thank you for the review. All suggestions are applied. Upgrading the green label.

**Pull request merged**